### PR TITLE
Adds billing address for Apple Pay

### PR DESCRIPTION
**Pull request by kamiranoff**

### CODE REVIEW

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 1]
```diff
@@ -332,15 +336,46 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    if (postalAddress.postalCode) [tmp setObject:postalAddress.postalCode forKey:@"postalCode"];
+    if (postalAddress.country) [tmp setObject:postalAddress.country forKey:@"country"];
+    if (postalAddress.ISOCountryCode) [tmp setObject:postalAddress.ISOCountryCode forKey:@"ISOCountryCode"];
+
+    NSError *error;
+    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:tmp
+                                                       options: 0
+                                                         error:&error];
+
+    if (!jsonData) {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: How about something like this instead?

if (error) {
    NSLog(@"Got an error: %@", error);
    return nil;
}

Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: should this be checking for `error` instead?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 2]
```diff
@@ -332,15 +336,46 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+        NSLog(@"Got an error: %@", error);
+    } else {
+        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+    }
+
+    return nil;
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: to clean this up a little, we should return `nil` when we know there is no data i.e. in the above conditional
as well as removing the `else`

Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Good point.

File: packages/react-native-payments/lib/js/PaymentRequest.js [hunk 3]
```diff
@@ -13,7 +13,7 @@ import type {
   PaymentAddress,
   PaymentShippingType,
   PaymentDetailsIOS,
-  PaymentDetailsIOSRaw,
+  PaymentDetailsIOSRaw
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/js/PaymentRequest.js]: seems your linter is code formatting this differently - could we keep it the same as per the library to avoid unnecessary changes like this (and the ones below)?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m (hunk 1 above)
Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Not sure..https://stackoverflow.com/questions/6368867/generate-json-string-from-nsdictionary-in-ios

File: packages/react-native-payments/lib/js/types.js [hunk 4]
```diff
@@ -86,14 +86,28 @@ export type PaymentShippingOption = {
 // https://www.w3.org/TR/payment-request/#paymentcomplete-enum
 export type PaymentComplete = 'fail' | 'success' | 'unknown';
 
+export type BillingAddress = {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/js/types.js]: Could any of these be nullable?

Comment (kamiranoff) [reply packages/react-native-payments/lib/js/types.js]: Good point. Added the `?` to the nullable ones.

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 5]
```diff
@@ -332,17 +336,102 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    if(billingContact.name.phoneticRepresentation.givenName) tmp[@"phoneticGivenName"] = billingContact.name.givenName;
+    
+    if (postalAddress.street) {
+        NSArray *streetArray = [postalAddress.street componentsSeparatedByString:@"\n"];
+        tmp[@"addressLines"] = streetArray;
+    }
+    if (postalAddress.city) tmp[@"locality"] = postalAddress.city;
+    if (postalAddress.subLocality) tmp[@"sublocality"] = postalAddress.subLocality;
+    if (postalAddress.subAdministrativeArea) tmp[@"subAdministrativeArea"] = postalAddress.subAdministrativeArea;
+    if (postalAddress.state) tmp[@"administrativeArea"] = postalAddress.state;
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: Just to confirm these mappings are the same as BPA Apple Pay right?

Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: yes, I have updated them to match BPA Apple Pay. Chris argued that should be done in the js side though not to impact too much the library.

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,4 @@
 node_modules/
 npm-debug.log
 lerna-debug.log
+.idea/
```
Comment (SukhKalsi) [inline .gitignore]: new line

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 7]
```diff
@@ -71,15 +71,24 @@ - (NSDictionary *)constantsToExport
...
 {
+    NSString * paymentStatus = [paymentObject valueForKey:@"status"];
+    
     if ([paymentStatus isEqualToString: @"success"]) {
         self.completion(PKPaymentAuthorizationStatusSuccess);
     } else {
-        self.completion(PKPaymentAuthorizationStatusFailure);
+        NSArray * errors = [paymentObject objectForKey:@"errors"];
+        for (id error in errors) {
+            if([[error valueForKey:@"error"]  isEqual: @"billingContactInvalid"]) {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: code formatting... space after `if` and double space before `isEqual`

File: .gitignore (hunk 6 above)
Comment (kamiranoff) [reply .gitignore]: 👍 done

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 8]
```diff
@@ -332,17 +350,183 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    }
+    
+    return result;
+}
+
+- (NSString *)getBillingContact:(PKContact *) billingContact {
+    NSMutableDictionary *tmp = [NSMutableDictionary dictionary];
+    CNPostalAddress *postalAddress = billingContact.postalAddress;
+    
+    if(billingContact.name.familyName) tmp[@"familyName"] = billingContact.name.familyName;
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: space after `if` (next few lines down same too :)

File: packages/react-native-payments/lib/ios/ReactNativePayments.m (hunk 5 above)
Comment (SukhKalsi) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Cool could we add a code reference linking to BPA code so we know what the mappings are in case they ever change in BPA?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 9]
```diff
@@ -332,17 +350,183 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    NSString *nameRegex = @"^[ ',-.0-9A-Za-zÀ-ÿ]*$";
+    NSPredicate *nameValidate = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", nameRegex];
+    
+    return [nameValidate evaluateWithObject:name];
+}
+
+
+-(BOOL)isValidPostCode:(NSString *) postCode {
+    
+    if (!postCode || ![self isValidName:postCode] || postCode.length > 10) {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: postcode has same validation as name?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 10]
```diff
@@ -332,17 +350,183 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    
+}
+
+-(BOOL)isValidStreet:(NSString *) street {
+    
+    NSArray* addressLines = [street componentsSeparatedByString: @"\n"];
+    NSString* line1 = [addressLines objectAtIndex: 0];
+    NSString* line2 = nil;
+    if([addressLines count] >= 2) {
+        line2 = [addressLines objectAtIndex: 1];
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: what happens to the remaining address if there are more than one in this array?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 11]
```diff
@@ -332,17 +350,183 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+    if(line2 && (![self isValidName:line2] || line2.length > 35)) {
+        return NO;
+    }
+    
+    return YES;
+}
+
+-(BOOL)isValidCityName:(NSString *) city {
+    
+    if(!city || ![self isValidName:city] || city.length > 20) {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: Seems like `isValidName` is used a lot :) maybe worth renaming the method to have `isValidString`

Comment (SukhKalsi) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Also why are we checking against string length? Can the backend not handle any strings that are longer gracefully?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m [hunk 12]
```diff
@@ -332,17 +350,183 @@ - (void)setRequiredShippingAddressFieldsFromOptions:(NSDictionary *_Nonnull)opti
...
+}
+
+
+-(BOOL)isValidBillingContact:(CNPostalAddress *) address {
+    
+    if(![self isValidPostCode:address.postalCode]) {
+        return NO;
+    }
+    
+    if(![self isValidStreet:address.street]) {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/ios/ReactNativePayments.m]: probably better to group this all in one conditional

File: packages/react-native-payments/lib/js/PaymentRequest.js [hunk 13]
```diff
@@ -383,7 +350,7 @@ export default class PaymentRequest {
   _removeEventListeners() {
     // Internal Events
     DeviceEventEmitter.removeSubscription(this._userDismissSubscription);
-    DeviceEventEmitter.removeSubscription(this._userAcceptSubscription);
+   // DeviceEventEmitter.removeSubscription(this._userAcceptSubscription);
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/js/PaymentRequest.js]: Should this be commented out and if so why? :)

File: packages/react-native-payments/lib/js/PaymentResponse.js [hunk 14]
```diff
@@ -48,8 +54,13 @@ export default class PaymentResponse {
   }
 
   // https://www.w3.org/TR/payment-request/#details-attribute
-  get details(): PaymentDetailsInit {
-    return this._details;
+  get billingContact(): any {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/js/PaymentResponse.js]: should this be `BillingContact` rather than `any` type that is returned?

File: packages/react-native-payments/lib/js/PaymentResponse.js [hunk 15]
```diff
@@ -48,8 +54,13 @@ export default class PaymentResponse {
...
+    return this._billingContact;
+  }
+
+  // https://www.w3.org/TR/payment-request/#details-attribute
+  get token(): any {
```
Comment (SukhKalsi) [inline packages/react-native-payments/lib/js/PaymentResponse.js]: Same here - is `any` right here?

File: packages/react-native-payments/lib/ios/ReactNativePayments.m (hunk 10 above)
Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Hum...There are only two lines for the address in the apple sheet.

File: packages/react-native-payments/lib/ios/ReactNativePayments.m (hunk 11 above)
Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: Yep. I agree I am a bit dubious about these validation rules. I'll add a link to where they point in BPA

Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: oh, and btw, these methods are not used anymore. Cause now we rely on BPA validation

File: packages/react-native-payments/lib/js/PaymentRequest.js (hunk 13 above)
Comment (kamiranoff) [reply packages/react-native-payments/lib/js/PaymentRequest.js]: I moved it inside of `app-discovery` inside of `utils/payment.js`

File: packages/react-native-payments/lib/ios/ReactNativePayments.m (hunk 5 above)
Comment (kamiranoff) [reply packages/react-native-payments/lib/ios/ReactNativePayments.m]: good point! I added a link
